### ISSUE TITLE
NSUrlSessionHandler allocations can impact performance due to high number of allocations

### ISSUE THREAD

**Issue by Therzok** [xamarin/xamarin-macios#8867]
### Steps to Reproduce

1. [DidReceiveResponse](https://gist.github.com/Therzok/93a448f0896cd3c73f073c97eca942c4)
2. [DidCompleteWithError](https://gist.github.com/Therzok/c89910ca05690fab6769d87f385679b0)
3. [DidReceiveResponse](https://gist.github.com/Therzok/8eacecbf03adc96b1bee373d529dbfae) again

Here are some sample traces taken while using VSMac and triggering nuget restore on the vsmac solution.

### Expected Behavior

⏩

**Comment by spouliot** [comment]
`xamarin_get_frame_length` is only hit depending on your configuration
Caching it has been tried but did not help (make it worse) as it's not that costly (compared to the caching cost)
There's more details inside https://github.com/xamarin/xamarin-macios/issues/3488

**Comment by Therzok** [comment]
Yeah, these are the CPU times, but there are still a large number of managed allocations (strings, arrays) coming in from marshalling.

**Comment by rolfbjarne** [comment]
> 1\. DidReceiveResponse

This should be fixed in #23825.

> 2\. DidCompleteWithError

The Dispose call here happens once per request, so I'm not sure if this is really that much of an issue.

> 3\. [DidReceiveResponse](https://gist.github.com/Therzok/8eacecbf03adc96b1bee373d529dbfae) again

This doesn't happen when building for Release.
